File: app/src/main/java/com/example/pc/oreo/WifiConnector.java
package com.example.pc.oreo;

import android.util.Log;

import com.example.pc.oreo.WifiCommand.WifiCommandCode;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.nio.channels.DatagramChannel;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import io.reactivex.FlowableOnSubscribe;
import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subscribers.DisposableSubscriber;

public class WifiConnector {
    private DatagramSocket datagramSocket;
    private DatagramSocket videoSocket;
    private InetAddress serverAddress;
    private int port;
    private TelloWifiDataChangeListener wifiDataChangeListener;
    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public WifiConnector(String ad, int serverPort) {
        try {
            serverAddress = InetAddress.getByName(ad);
            port = serverPort;
            datagramSocket = new DatagramSocket();
            videoSocket = new DatagramSocket();
        } catch (SocketException e) {
            datagramSocket = null;
            videoSocket = null;
        } catch (UnknownHostException e) {
            serverAddress = null;
        }
    }

    private byte[] formRequestPacket(int port) {
        byte[] connectPacket = new byte[64];
        connectPacket[0] = WifiCommand.WifiCommandCode.REQUEST_CONNECTION.getLength();
        connectPacket[1] = WifiCommand.WifiCommandCode.REQUEST_CONNECTION.getValue();
        return connectPacket;
    }

    public void connect() {
        compositeDisposable.add(
                Observable.create(new ObservableOnSubscribe<Integer>() {
                    @Override
                    public void subscribe(ObservableEmitter<Integer> emitter) {
                        if (emitter.isDisposed()) {
                            return;
                        }
                        try {
                            DatagramChannel channel = DatagramChannel.open();
                            datagramSocket = channel.socket();
                            datagramSocket.connect(serverAddress, port);
                            byte[] connectPacket = formRequestPacket(videoSocket.getLocalPort());
                            sendWithException(connectPacket);
                            byte[] buffer = new byte[2048];
                            DatagramPacket pk = new DatagramPacket(buffer, buffer.length);
                            datagramSocket.setSoTimeout(100);
                            datagramSocket.receive(pk);
                            int replyCode = pk.getData()[1];
                            if (replyCode == WifiCommandCode.APPROVE_CONNECTION.getValue()) {
                                emitter.onNext(replyCode);
                                emitter.onComplete();
                            } else {
                                emitter.onError(new Throwable());
                            }
                        } catch (IOException e) {
                            if (!emitter.isDisposed()) {
                                emitter.onError(e);
                            }
                        }
                    }
                }).subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableObserver<Integer>() {
                            @Override
                            public void onNext(Integer replyCode) {
                                Log.e("connection", replyCode.toString());
                                wifiDataChangeListener.onWifiCommandReceived(new WifiCommand(replyCode));
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("connectionError", "AAAAAAAAAA");
                            }

                            @Override
                            public void onComplete() {
                                Log.e("connectionComplete", "BBBBBBBB");
                                streamData();
                                dispose();
                            }
                        })
        );
    }

    public void send(final byte[] message) {
        compositeDisposable.add(
                Observable.create(new ObservableOnSubscribe<Boolean>() {
                    @Override
                    public void subscribe(ObservableEmitter<Boolean> emitter) {
                        if (emitter.isDisposed()) {
                            return;
                        }
                        try {
                            DatagramPacket datagramPacket = new DatagramPacket(message, message.length, serverAddress, port);
                            datagramSocket.send(datagramPacket);
                            emitter.onNext(true);
                            emitter.onComplete();
                        } catch (IOException e) {
                            if (!emitter.isDisposed()) {
                                emitter.onError(e);
                            }
                        }
                    }
                }).subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .retry()
                        .subscribeWith(new DisposableObserver<Boolean>() {
                            @Override
                            public void onNext(Boolean sent) {
                                Log.e("sending", "");
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("sendingError", "AAAAAAAAAA");
                            }

                            @Override
                            public void onComplete() {
                                Log.e("sendingComplete", "BBBBBBBB");
                                dispose();
                            }
                        })
        );
    }

    private void streamData() {
        compositeDisposable.add(
                Flowable.create(new FlowableOnSubscribe<DatagramPacket>() {
                    @Override
                    public void subscribe(FlowableEmitter<DatagramPacket> emitter) throws Exception {
                        byte[] buffer = new byte[1024];
                        DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                        datagramSocket.setSoTimeout(150);
                        while (true) {
                            if (emitter.isCancelled()) {
                                break;
                            }
                            try {
                                if (!datagramSocket.isClosed()) {
                                    datagramSocket.receive(packet);
                                    emitter.onNext(packet);
                                } else {
                                    Log.e("socket closed", "");
                                    emitter.onError(new Throwable());
                                    break;
                                }
                            } catch (IOException e) {
                                Log.e("UDP packet error", "");
                                if (!emitter.isCancelled()) {
                                    emitter.onError(e);
                                }
                                break;
                            }
                        }
                    }
                }, BackpressureStrategy.LATEST)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.trampoline())
                        .retry()
                        .subscribeWith(new DisposableSubscriber<DatagramPacket>() {
                            @Override
                            public void onNext(DatagramPacket datagramPacket) {
                                Log.e("receive ", datagramPacket.getData().length + "");
                                WifiCommand command = new WifiCommand(datagramPacket.getData());
                                wifiDataChangeListener.onWifiCommandReceived(command);
                                Log.e("receiveNext", command.getType() + "");
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("receiveError", "AAAAAAAAAA");
                                dispose();
                            }

                            @Override
                            public void onComplete() {
                                Log.e("receiveComplete", "BBBBBBBB");
                                dispose();
                            }
                        })
        );
    }

    private void startStreaming() {
        compositeDisposable.add(
                Flowable.create(new FlowableOnSubscribe<DatagramPacket>() {
                    @Override
                    public void subscribe(FlowableEmitter<DatagramPacket> emitter) throws Exception {
                        byte[] buffer = new byte[2048];
                        videoSocket.setSoTimeout(100);
                        DatagramPacket packet;
                        while (true) {
                            if (emitter.isCancelled()) {
                                break;
                            }
                            try {
                                if (!videoSocket.isClosed()) {
                                    packet = new DatagramPacket(buffer, buffer.length);
                                    videoSocket.receive(packet);
                                    emitter.onNext(packet);
                                } else {
                                    Log.e("video socket closed", "");
                                    emitter.onError(new Throwable());
                                    break;
                                }
                            } catch (SocketTimeoutException e) {
                                continue;
                            } catch (IOException e) {
                                Log.e("video packet error", "");
                                if (!emitter.isCancelled()) {
                                    emitter.onError(e);
                                }
                            }
                        }
                    }
                }, BackpressureStrategy.BUFFER)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.trampoline())
                        .subscribeWith(new DisposableSubscriber<DatagramPacket>() {
                            @Override
                            public void onNext(DatagramPacket packet) {
                                Log.e("streamNext", packet.getLength() + "");
                                wifiDataChangeListener.onVideoDataReceived(packet.getData(), packet.getLength());
                            }

                            @Override
                            public void onError(Throwable e) {
                                Log.e("streamError", "AAAAAAAAAA");
                            }

                            @Override
                            public void onComplete() {
                                Log.e("streamComplete", "BBBBBBBB");
                            }
                        })
        );
    }

    private void sendWithException(byte[] message) throws IOException {
        DatagramPacket datagramPacket = new DatagramPacket(message, message.length, serverAddress, port);
        datagramSocket.send(datagramPacket);
    }

    public void disconnect() {
        if (!compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
        if (datagramSocket != null) {
            datagramSocket.close();
        }
        if (videoSocket != null) {
            videoSocket.close();
        }
    }

    public DatagramSocket getDatagramSocket() {
        return datagramSocket;
    }

    public DatagramSocket getVideoSocket() {
        return videoSocket;
    }

    public void setWifiDataChangeListener(TelloWifiDataChangeListener wifiDataChangeListener) {
        this.wifiDataChangeListener = wifiDataChangeListener;
    }

    public interface TelloWifiDataChangeListener {
        void onWifiCommandReceived(WifiCommand command);

        void onVideoDataReceived(byte[] videoBuffer, int size);
    }
}

File: app/src/main/java/com/example/pc/oreo/WifiConnector.kt
package com.example.pc.oreo

import android.util.Log
import io.reactivex.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableObserver
import io.reactivex.schedulers.Schedulers
import io.reactivex.subscribers.DisposableSubscriber
import java.io.IOException
import java.net.*
import java.nio.channels.DatagramChannel
import java.util.concurrent.TimeUnit
import kotlin.text.Charsets.UTF_8

class WifiConnector(ad: String, serverPort: Int) {
    var datagramSocket: DatagramSocket? = null
        private set
    private var serverAddress: InetAddress? = null
    private var port: Int = 0
    var wifiDataChangeListener: OreoWifiDataChangeListener? = null
    private val compositeDisposable = CompositeDisposable()
    lateinit var oreo: OreoCommandListener
    private val bufferSize = 128

    init {
        try {
            serverAddress = InetAddress.getByName(ad)
            port = serverPort
        } catch (e: UnknownHostException) {
            serverAddress = null
        }
    }

    private fun formRequestPacket(): ByteArray {
        val connectPacket = ByteArray(1)
        connectPacket[0] = WifiCommandCode.REQUEST_CONNECTION.value
        return connectPacket
    }

    fun connect() {
        compositeDisposable.add(
                Observable.create(ObservableOnSubscribe<DatagramPacket> { emitter ->
                    if (emitter.isDisposed) {
                        return@ObservableOnSubscribe
                    }
                    try {
                        datagramSocket = DatagramChannel.open().socket()
                        datagramSocket?.apply {
                            connect(serverAddress,port)
                            val connectPacket = formRequestPacket()
                            val datagramPacket = DatagramPacket(connectPacket, connectPacket.size, serverAddress, port)
                            send(datagramPacket)
                            val buffer = ByteArray(bufferSize)
                            val pk = DatagramPacket(buffer, buffer.size)
                            soTimeout = 100
                            receive(pk)
                            emitter.onNext(pk)
                        }
                    } catch (e: IOException) {
                        if (!emitter.isDisposed) {
                            emitter.onError(e)
                        }
                    }
                }).subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableObserver<DatagramPacket>() {
                            override fun onNext(datagramPacket: DatagramPacket) {
                                Log.e("connection", "")
                                wifiDataChangeListener?.onDriveDataReceived(datagramPacket.data.toString(UTF_8))
                            }

                            override fun onError(e: Throwable) {
                                Log.e("connectionError", "AAAAAAAAAA")
                            }

                            override fun onComplete() {
                                Log.e("connectionComplete", "BBBBBBBB")
                                streamData()
                                dispose()
                            }
                        })
        )
    }

    fun streamControls() {
        compositeDisposable.add(
                Flowable.interval(100, TimeUnit.MILLISECONDS)
                        .onBackpressureLatest()
                        .map {
                            val controls = oreo.onNotifyCommandUpdate()
                            val datagramPacket = DatagramPacket(controls, controls.size, serverAddress, port)
                            datagramSocket?.soTimeout = 150
                            datagramSocket?.send(datagramPacket)
                            true
                        }
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.trampoline())
                        .doOnError { err -> Log.e("control send error", err.localizedMessage) }
                        .retry()
                        .subscribeWith(object : DisposableSubscriber<Boolean>() {
                            override fun onNext(result: Boolean) {
                                Log.e("control sent ", result.toString())
                            }

                            override fun onError(e: Throwable) {
                                Log.e("sendError", e.localizedMessage)
                                dispose()
                            }

                            override fun onComplete() {
                                Log.e("sendComplete", "BBBBBBBB")
                                dispose()
                            }
                        })
        )
    }

    private fun streamData() {
        compositeDisposable.add(
                Flowable.create(FlowableOnSubscribe<DatagramPacket> { emitter ->
                    val buffer = ByteArray(bufferSize)
                    val packet = DatagramPacket(buffer, buffer.size)
                    datagramSocket?.soTimeout = 150
                    while (true) {
                        if (emitter.isCancelled) {
                            break
                        }
                        try {
                            if (!datagramSocket?.isClosed!!) {
                                datagramSocket?.receive(packet)
                                emitter.onNext(packet)
                            } else {
                                Log.e("socket closed", "")
                                emitter.onError(Throwable())
                                break
                            }
                        } catch (e: IOException) {
                            Log.e("UDP packet error", "")
                            if (!emitter.isCancelled) {
                                emitter.onError(e)
                            }
                            break
                        }
                    }
                }, BackpressureStrategy.LATEST)
                        .subscribeOn(Schedulers.io())
                        .observeOn(Schedulers.trampoline())
                        .retry()
                        .subscribeWith(object : DisposableSubscriber<DatagramPacket>() {
                            override fun onNext(datagramPacket: DatagramPacket) {
                                Log.e("receive ", datagramPacket.data.size.toString() + "")
                                wifiDataChangeListener?.onDriveDataReceived(datagramPacket.data.toString(UTF_8))
                            }

                            override fun onError(e: Throwable) {
                                Log.e("receiveError", "AAAAAAAAAA")
                                dispose()
                            }

                            override fun onComplete() {
                                Log.e("receiveComplete", "BBBBBBBB")
                                dispose()
                            }
                        })
        )
    }

    fun disconnect() {
        if (!compositeDisposable.isDisposed) {
            compositeDisposable.dispose()
        }
        datagramSocket?.close()
    }

    interface OreoWifiDataChangeListener {
        fun onDriveDataReceived(jsonData: String)

        fun onVideoDataReceived(videoBuffer: ByteArray, size: Int)
    }

    interface OreoCommandListener {
        fun onNotifyCommandUpdate(): ByteArray
    }

    enum class WifiCommandCode constructor(val value: Byte) {
        REQUEST_CONNECTION(100),
        APPROVE_CONNECTION(101),
        JOYSTICK_CONTROL(102),
        REQUEST_STREAM(105),
        FRAME_DATA(106),
        DRIVE_DATA(107);
    }
}
